wp/lib/bap_wp/src/constraint.ml

```diff
@@ -183,25 +190,108 @@ type t =
+(* preen_expr expr will improve the readability
+   of expr by getting rid of white space 
+   and simplifying the expression. *)
+let preen_expr (expr : Expr.expr) : string =
+  Expr.simplify expr None
+  |> Expr.to_string
+  |> String.substr_replace_all ~pattern:"  " ~with_:""
+
+(* delete_empty_constr_hyps const will remove 
+   some empty lists in Clause constructs. For example
+   Clause([],Clause[Clause([],[x]),y]) becomes 
+   Clause([],Clause[x,y]). *)
+let rec del_empty_constr_hyps (const : t) : t =
+  match const with
+  | Clause(hyps,concs) ->
+    let fold_list c_list =
+      List.fold_right c_list ~init:[]
+        ~f:(fun x y ->
+          match x with
+          | Clause([],xs) ->
+             List.append (List.map xs ~f:del_empty_constr_hyps) y
+          | _ -> (del_empty_constr_hyps x) :: y) in
+    Clause(fold_list hyps, fold_list concs)
+  | Goal g -> Goal g
+  | ITE (ite, e, c1, c2) ->
+     let new_c1 = del_empty_constr_hyps c1 in
+     let new_c2 = del_empty_constr_hyps c2 in
+     ITE (ite, e, new_c1, new_c2)
   | Subst (c, olds, news) ->
-    Format.fprintf ch "Substitute: %s to %s in %a"
-      (List.to_string ~f:Expr.to_string olds) (List.to_string ~f:Expr.to_string news)
-      pp_constr c
-
-let to_string (constr : t) : string =
-  Format.asprintf "%a" pp_constr constr
+     let new_c = del_empty_constr_hyps c in
+     Subst (new_c, olds, news)
+
+(* to_color colorful ch c str ins will print str with insert
+   ins and formatter ch. If colorful is true, then this will 
+   print in the color c. *)
+let to_color (colorful : bool) (ch : Format.formatter) (c : int)
+      (str : ('a -> 'b, Format.formatter, unit) Stdlib.format)
+      (ins : string) : unit =
+    if colorful then Format.fprintf ch "\x1b[%dm" c;
+    Format.fprintf ch str ins;
+    if colorful then Format.fprintf ch "\x1b[0m"
+
+(* print_expr_list expr_list ch converts expr_list into a list of 
+   strings of expressions and then prints each one on a new line 
+   using the formatter ch *)
+let print_expr_list (expr_list : Expr.expr list)
+      (ch : Format.formatter) : unit =
+  let expr_string_list : string list =
+    List.map expr_list ~f:(fun e -> preen_expr e |> String.split_lines)
+    |> List.join in
+  let rec print_exprs ch exprs =
+    match exprs with
+    | [] -> ();
+    | [e] -> Format.fprintf ch "%s" e
+    | e :: es -> Format.fprintf ch "%s@;" e ; print_exprs ch es in 
+  Format.fprintf ch "@[<v 2>%a@]" print_exprs expr_string_list
+
+let pp_constr (colorful : bool) (form : Format.formatter)
+      (cons : t) : unit =
+  let rec rec_pp_constr ch constr =
+    let to_color = to_color colorful ch in
+    match constr with
+    | Goal g ->
+       Format.fprintf ch "%s@;" (goal_to_string ~colorful:colorful g)
+    | ITE (tid, e, c1, c2) ->
+       let color = 35 in (* green *)
+       to_color color "%s: (if " (tid |> Term.tid |> Tid.to_string);
+       Format.fprintf ch "%s" (preen_expr e);
+       to_color color " then%s" "";
```

Comment:
This is an ugly hack and should be changed. Ideally `to_color` should just take in a color and a format-string, and then the number of inputs for the format-string wouldn't be fixed. The way I've designed it here is that the format-string takes in exactly one input. Thus we end up with these cases where we have `" %s" ""` which are ugly.